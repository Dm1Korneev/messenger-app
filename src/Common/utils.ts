
type Filter = Parameters<typeof Array.prototype.filter>[0]
export const distinct: Filter = (value, index, self) => self.indexOf(value) === index;
