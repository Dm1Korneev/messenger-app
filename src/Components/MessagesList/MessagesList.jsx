import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import withStyles from '@material-ui/core/styles/withStyles';
import List from '@material-ui/core/List';
import Divider from '@material-ui/core/Divider';
import Box from '@material-ui/core/Box';

import MessageDateTime from 'Components/MessageDateTime';
import MessageText from 'Components/MessageText';
import MessageUser from 'Components/MessageUser';
import { RELOAD_PERIOD } from 'Constants';

class MessagesList extends React.Component {
  messagesEnd = React.createRef();

  constructor(props) {
    super(props);

    this.loadMessages = props.loadMessages;
    this.interval = undefined;
    this.state = { bottomPosition: true };
  }

  componentDidMount() {
    this.scrollToBottom();
    this.interval = setInterval(this.loadMessages, RELOAD_PERIOD);
  }

  componentDidUpdate() {
    const { bottomPosition } = this.state;
    if (bottomPosition) {
      this.scrollToBottom();
    }
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  scrollToBottom = () => {
    if (this.messagesEnd.current) {
      this.messagesEnd.current.scrollIntoView({
        blok: 'end',
        behavior: 'smooth',
      });
    }
  };

  handlerOnScroll = (event) => {
    const { bottomPosition } = this.state;
    const newBottomPosition = event.target.scrollHeight - event.target.offsetHeight
      === event.target.scrollTop;
    if (bottomPosition !== newBottomPosition) {
      this.setState({ bottomPosition: newBottomPosition });
    }
  };

  render() {
    const {
      messagesTree, classes, user, users,
    } = this.props;

    return (
      <List className={classes.list} onScroll={this.handlerOnScroll}>
        {messagesTree.map((valueAuthor, indexAuthor) => {
          const { author, childrens: childrensAuthor } = valueAuthor;

          const isCurrentUserMessage = user._id === author;
          const firstDateTime = childrensAuthor[0].dateTime;

          const childrenComponentsAuthor = childrensAuthor.map(
            (valueDateTime) => {
              const { dateTime, childrens: childrensDateTime } = valueDateTime;

              const childrenComponentsDateTime = childrensDateTime.map(
                (value) => (
                  <MessageText
                    key={value._id}
                    text={value.text}
                    isCurrentUserMessage={isCurrentUserMessage}
                  />
                ),
              );

              return (
                <MessageDateTime
                  key={`${author}-${dateTime}`}
                  dateTime={dateTime}
                  isCurrentUserMessage={isCurrentUserMessage}
                >
                  <Box display="flex" flexDirection="column">{childrenComponentsDateTime}</Box>
                </MessageDateTime>
              );
            },
          );

          const messageAuthor = users[author];

          let name;
          let avatar;
          if (messageAuthor) {
            name = messageAuthor.name;
            avatar = messageAuthor.avatar;
          }

          return (
            <Fragment key={`${author}-${firstDateTime}`}>
              {indexAuthor > 0 && (
                <Divider variant="middle" />
              )}
              <MessageUser
                author={name}
                avatar={avatar}
                isCurrentUserMessage={isCurrentUserMessage}
              >
                <Box display="flex" flexDirection="column">{childrenComponentsAuthor}</Box>
              </MessageUser>
            </Fragment>
          );
        })}
        <div ref={this.messagesEnd} />
      </List>
    );
  }
}

MessagesList.defaultProps = {
  messagesTree: [],
};
MessagesList.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  loadMessages: PropTypes.func.isRequired,
  messagesTree: PropTypes.arrayOf(PropTypes.object),
  user: PropTypes.shape({
    _id: PropTypes.string.isRequired,
  }).isRequired,
  users: PropTypes.instanceOf(Object).isRequired,
};

const styles = (theme) => ({
  list: {
    overflow: 'auto',
    flexGrow: 1,
    padding: theme.spacing(0),
  },
});

export default withStyles(styles)(MessagesList);

