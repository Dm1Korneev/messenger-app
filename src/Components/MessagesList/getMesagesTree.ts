
import { Message, User } from 'Types';

type DateTimeNode = {
  dateTime: Date;
  children: Message[];
}

type AuthorNode = {
  author: string;
  children: DateTimeNode[];
}

type MessagesTree = AuthorNode[]

export const getMesagesTree = ({ messages }: { messages: Message[], users: User[] }) => {
  const messagesTree: MessagesTree = [];
  let authorNode: AuthorNode;
  let dateTimeNode: DateTimeNode;
  let dateTime: Date;
  let lastNodeDateTime: Date;
  messages.forEach((element, index) => {
    dateTime = new Date(element.dateTime);
    const currentElementDateTime = new Date(dateTime.getTime());
    currentElementDateTime.setSeconds(0);
    currentElementDateTime.setMilliseconds(0);
    if (index === 0 || authorNode.author !== element.author) {
      messagesTree.push({ author: element.author, children: [] });
      authorNode = messagesTree[messagesTree.length - 1];
      authorNode.children.push({
        dateTime,
        children: [],
      });
      dateTimeNode = authorNode.children[authorNode.children.length - 1];
      dateTimeNode.children.push(element);
      lastNodeDateTime = currentElementDateTime;
    } else if (currentElementDateTime > lastNodeDateTime) {
      authorNode.children.push({
        dateTime,
        children: [],
      });
      dateTimeNode = authorNode.children[authorNode.children.length - 1];
      dateTimeNode.children.push(element);
      lastNodeDateTime = currentElementDateTime;
    } else {
      dateTimeNode.children.push(element);
    }
  });

  return messagesTree;
};
