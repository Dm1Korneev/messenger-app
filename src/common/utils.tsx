export function distinct(value: any, index: number, self: Array<any>): boolean {
  return self.indexOf(value) === index;
}
