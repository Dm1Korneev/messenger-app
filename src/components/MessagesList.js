import React, { Component } from "react";
import Message from "./Message";
import List from "@material-ui/core/List";
import withStyles from "@material-ui/core/styles/withStyles";

const styles = theme => ({
  list: {
    overflow: "auto",
    padding: 0
  }
});

class MessagesList extends Component {
  messagesEnd = React.createRef();

  scrollToBottom = () => {
    this.messagesEnd.current.scrollIntoView({
      blok: "end",
      behavior: "smooth"
    });
  };

  componentDidUpdate() {
    setTimeout(() => {
      this.scrollToBottom();
    }, 300);
  }

  render() {
    const { messages, classes } = this.props;
    return (
      <List className={classes.list}>
        {messages.map(value => (
          <Message key={value._id} message={value} />
        ))}
        <div ref={this.messagesEnd} />
      </List>
    );
  }
}

export default withStyles(styles)(MessagesList);
