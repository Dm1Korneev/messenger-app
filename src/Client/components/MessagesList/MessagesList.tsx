import Box from '@mui/material/Box';
import Divider from '@mui/material/Divider';
import List from '@mui/material/List';
import {
  Fragment, useEffect, useRef, useState,
} from 'react';

import { RELOAD_PERIOD } from 'Client/constants';
import { useUsers, useCurrentUser, useMessagesByChatId } from 'Client/hooks';

import { MessageDateTime } from '../MessageDateTime';
import { MessageText } from '../MessageText';
import { MessageUser } from '../MessageUser';

import { getMesagesTree } from './getMesagesTree';

type MessagesListProps = {
  activeChatId?: string
}

export const MessagesList = ({ activeChatId }: MessagesListProps) => {
  const { data: messages } = useMessagesByChatId(activeChatId, { refetchInterval: RELOAD_PERIOD });
  const messagesTree = getMesagesTree(messages);
  const { data: users } = useUsers();
  const { data: currentUser } = useCurrentUser();

  const messagesEnd = useRef<HTMLDivElement>(null);

  const [bottomPosition, setBottomPosition] = useState(true);

  const scrollToBottom = () => {
    if (messagesEnd.current) {
      messagesEnd.current.scrollIntoView({
        block: 'end',
        behavior: 'smooth',
      });
    }
  };

  useEffect(() => {
    scrollToBottom();
  }, [messagesTree]);

  const handlerOnScroll = (event: React.UIEvent<HTMLUListElement>) => {
    const newBottomPosition = event.currentTarget.scrollHeight - event.currentTarget.offsetHeight
      === event.currentTarget.scrollTop;
    if (bottomPosition !== newBottomPosition) {
      setBottomPosition(newBottomPosition);
    }
  };

  return (
    <List onScroll={handlerOnScroll} sx={{ flexGrow: 1, p: 0, overflow: 'auto' }}>
      {messagesTree.map((valueAuthor, indexAuthor) => {
        const { author, children: childrenAuthor } = valueAuthor;

        const isCurrentUserMessage = currentUser?._id === author;
        const firstDateTime = childrenAuthor[0].dateTime;

        const childrenComponentsAuthor = childrenAuthor.map(
          (valueDateTime) => {
            const { dateTime, children: childrenDateTime } = valueDateTime;

            const childrenComponentsDateTime = childrenDateTime.map(
              (value) => (
                <MessageText
                  key={value._id}
                  isCurrentUserMessage={isCurrentUserMessage}
                  text={value.text}
                />
              ),
            );

            return (
              <MessageDateTime
                key={`${author}-${dateTime}`}
                dateTime={dateTime}
                isCurrentUserMessage={isCurrentUserMessage}
              >
                <Box display="flex" flexDirection="column">{childrenComponentsDateTime}</Box>
              </MessageDateTime>
            );
          },
        );

        const messageAuthor = users.find(({ _id }) => author === _id);

        return (
          <Fragment key={`${author}-${firstDateTime}`}>
            {indexAuthor > 0 && (
            <Divider variant="middle" />
            )}
            <MessageUser
              author={messageAuthor?.name ?? ''}
              avatar={messageAuthor?.avatar ?? ''}
              isCurrentUserMessage={isCurrentUserMessage}
            >
              <Box display="flex" flexDirection="column">{childrenComponentsAuthor}</Box>
            </MessageUser>
          </Fragment>
        );
      })}
      <div ref={messagesEnd} />
    </List>
  );
};
