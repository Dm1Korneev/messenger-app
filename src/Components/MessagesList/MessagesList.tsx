import Box from '@material-ui/core/Box';
import Divider from '@material-ui/core/Divider';
import List from '@material-ui/core/List';
import { makeStyles } from '@material-ui/core/styles';
import {
  Fragment, useEffect, useRef, useState,
} from 'react';

import { MessageDateTime } from 'Components/MessageDateTime';
import { MessageText } from 'Components/MessageText';
import { MessageUser } from 'Components/MessageUser';
import { RELOAD_PERIOD } from 'Constants';
import { useUsers, useCurrentUser, useMessagesByChatId } from 'Hooks';

import { getMesagesTree } from './getMesagesTree';

const useStyles = makeStyles((theme) => ({
  list: {
    overflow: 'auto',
    flexGrow: 1,
    padding: theme.spacing(0),
  },
}));

type MessagesListProps = {
  activeChatId?: string
}

export const MessagesList = ({ activeChatId }: MessagesListProps) => {
  const { data: messages } = useMessagesByChatId(activeChatId, { refetchInterval: RELOAD_PERIOD });
  const messagesTree = getMesagesTree(messages);
  const { data: users } = useUsers();
  const { data: currentUser } = useCurrentUser();

  const classes = useStyles();
  const messagesEnd = useRef<HTMLDivElement>(null);

  const [bottomPosition, setBottomPosition] = useState(true);

  const scrollToBottom = () => {
    if (messagesEnd.current) {
      messagesEnd.current.scrollIntoView({
        block: 'end',
        behavior: 'smooth',
      });
    }
  };

  useEffect(() => {
    scrollToBottom();
  }, [messagesTree]);

  const handlerOnScroll = (event: React.UIEvent<HTMLUListElement>) => {
    const newBottomPosition = event.currentTarget.scrollHeight - event.currentTarget.offsetHeight
      === event.currentTarget.scrollTop;
    if (bottomPosition !== newBottomPosition) {
      setBottomPosition(newBottomPosition);
    }
  };

  return (
    <List className={classes.list} onScroll={handlerOnScroll}>
      {messagesTree.map((valueAuthor, indexAuthor) => {
        const { author, children: childrenAuthor } = valueAuthor;

        const isCurrentUserMessage = currentUser?._id === author;
        const firstDateTime = childrenAuthor[0].dateTime;

        const childrenComponentsAuthor = childrenAuthor.map(
          (valueDateTime) => {
            const { dateTime, children: childrenDateTime } = valueDateTime;

            const childrenComponentsDateTime = childrenDateTime.map(
              (value) => (
                <MessageText
                  key={value._id}
                  text={value.text}
                  isCurrentUserMessage={isCurrentUserMessage}
                />
              ),
            );

            return (
              <MessageDateTime
                key={`${author}-${dateTime}`}
                dateTime={dateTime}
                isCurrentUserMessage={isCurrentUserMessage}
              >
                <Box display="flex" flexDirection="column">{childrenComponentsDateTime}</Box>
              </MessageDateTime>
            );
          },
        );

        const messageAuthor = users.find(({ _id }) => author === _id);

        return (
          <Fragment key={`${author}-${firstDateTime}`}>
            {indexAuthor > 0 && (
            <Divider variant="middle" />
            )}
            <MessageUser
              author={messageAuthor?.name ?? ''}
              avatar={messageAuthor?.avatar ?? ''}
              isCurrentUserMessage={isCurrentUserMessage}
            >
              <Box display="flex" flexDirection="column">{childrenComponentsAuthor}</Box>
            </MessageUser>
          </Fragment>
        );
      })}
      <div ref={messagesEnd} />
    </List>
  );
};
