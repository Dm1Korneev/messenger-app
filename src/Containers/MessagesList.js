import commonHoc from 'Containers/commonHoc';

import { loadMessages } from 'Redux/actions';

import { usersByIdSelector } from 'Selectors/users';
import { currentUserSelector } from 'Selectors/session';
import { messagesTreeSelector } from 'Selectors/messages';

import MessagesList from 'Components/MessagesList';

const mapStateToProps = (state) => ({
  messagesTree: messagesTreeSelector(state),
  users: usersByIdSelector(state),
  user: currentUserSelector(state),
});

const mapDispatchToProps = {
  loadMessages,
};

export default commonHoc(MessagesList, {
  mapStateToProps,
  mapDispatchToProps,
});
