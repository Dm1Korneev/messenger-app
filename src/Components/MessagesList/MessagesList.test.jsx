import React from 'react';
import { render } from '@testing-library/react';

import { mockComponent } from 'testing/utils';

import MessagesList from './MessagesList';

jest.mock('Components/MessageUser', () => mockComponent('MessageUser'));
jest.mock('Components/MessageDateTime', () => mockComponent('MessageDateTime'));
jest.mock('Components/MessageText', () => mockComponent('MessageText'));

jest.mock('@material-ui/core/List', () => mockComponent('List'));
jest.mock('@material-ui/core/Divider', () => mockComponent('Divider'));

const scrollIntoViewMock = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

const props = {
  loadMessages: jest.fn(),
  user: {
    _id: '',
  },
  users: [],
};

describe('render MessagesList component', () => {
  const setup = () => render(
    <MessagesList {...props} />,
  );

  test('snapshot test', async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  test('Scroll to and of messages is called on mout', () => {
    setup();

    expect(scrollIntoViewMock).toHaveBeenCalled();
  });
});

