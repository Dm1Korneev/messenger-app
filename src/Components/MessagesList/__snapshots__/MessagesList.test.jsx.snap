// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MessagesList component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="List"
    test-prop-classname="\\"MessagesList-list-1\\""
  >
    <div />
  </div>
</DocumentFragment>
`;
